### PR TITLE
rpz dumper: stop generating double zz labels on networks that start with zeroes

### PR DESCRIPTION
**Pull request by Habbie**
### Short description
Includes extra tests from @rgacogne and @omoerbeek.

This may not be the best fix - perhaps 'are we at the beginning of the string' is a better one. Perhaps spending some more time with this whole function is appropriate.

I also suspect we get this bit wrong:

>  If there exists more than one sequence of zero-valued fields of identical length, then only the last such sequence is compressed. 

### Checklist

I have:
- [ ] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Comment by Habbie** [comment]
(also, I left some debugging in)

File: pdns/recursordist/test-filterpo_cc.cc [hunk 1]
```diff
@@ -814,3 +814,8 @@ BOOST_AUTO_TEST_CASE(test_multiple_filter_policies_order)
     BOOST_CHECK(matchingPolicy.d_kind == DNSFilterEngine::PolicyKind::NoAction);
   }
 }
+
+BOOST_AUTO_TEST_CASE(test_mask_to_rpz)
+{
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("::2/127")).toString(), "127.2.zz.");
```
Comment (omoerbeek) [inline pdns/recursordist/test-filterpo_cc.cc:820]:
```suggestion
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("::2/127")).toString(), "127.2.zz.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1::2/127")).toString(), "127.2.zz.1.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("2::/127")).toString(), "127.zz.2.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1abc:2::/127")).toString(), "127.zz.2.1abc.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:2:3:4:5:6:7::/127")).toString(), "127.0.7.6.5.4.3.2.1.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:2:3:4:5:6::/127")).toString(), "127.zz.6.5.4.3.2.1.");
```

Comment (Habbie) [reply pdns/recursordist/test-filterpo_cc.cc:820]: thanks, these all pass!

**Comment by Habbie** [comment]
> I also suspect we get this bit wrong:
> 
> > If there exists more than one sequence of zero-valued fields of identical length, then only the last such sequence is compressed.

Yes, we get this wrong. Will fix.

**Comment by Habbie** [comment]
I added a test for 'more than one-sequence of the same length' (which failed) and then I rewrote the whole function. Now all tests pass.

**Comment by Habbie** [comment]
squashed; marked ready for review

File: pdns/recursordist/test-filterpo_cc.cc [hunk 2]
```diff
@@ -814,3 +814,14 @@ BOOST_AUTO_TEST_CASE(test_multiple_filter_policies_order)
...
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1::2/127")).toString(), "127.2.zz.1.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("2::/127")).toString(), "127.zz.2.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1abc:2::/127")).toString(), "127.zz.2.1abc.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:2:3:4:5:6:7::/127")).toString(), "127.0.7.6.5.4.3.2.1.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:2:3:4:5:6::/127")).toString(), "127.zz.6.5.4.3.2.1.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:0:2:0:0:0/127")).toString(), "127.zz.2.0.0.0.1.");
+}
```
Comment (omoerbeek) [inline pdns/recursordist/test-filterpo_cc.cc:827]:
```suggestion
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:2:0:0:0:0/127")).toString(), "127.zz.2.0.0.1.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:0:0:2:0:0/127")).toString(), "127.0.0.2.zz.1.");
}
```

File: pdns/recursordist/test-filterpo_cc.cc [hunk 3]
```diff
@@ -814,3 +814,14 @@ BOOST_AUTO_TEST_CASE(test_multiple_filter_policies_order)
...
+
+BOOST_AUTO_TEST_CASE(test_mask_to_rpz)
+{
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("::2/127")).toString(), "127.2.zz.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1::2/127")).toString(), "127.2.zz.1.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("2::/127")).toString(), "127.zz.2.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1abc:2::/127")).toString(), "127.zz.2.1abc.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:2:3:4:5:6:7::/127")).toString(), "127.0.7.6.5.4.3.2.1.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:2:3:4:5:6::/127")).toString(), "127.zz.6.5.4.3.2.1.");
+  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:0:2:0:0:0/127")).toString(), "127.zz.2.0.0.0.1.");
```
Comment (omoerbeek) [inline pdns/recursordist/test-filterpo_cc.cc:826]:
```suggestion
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:0:2:0:0:0/127")).toString(), "127.zz.2.0.0.0.1.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:2:0:0:0:0/127")).toString(), "127.zz.2.0.0.1.");
  BOOST_CHECK_EQUAL(DNSFilterEngine::Zone::maskToRPZ(Netmask("1:0:0:0:0:2:0:0/127")).toString(), "127.0.0.2.zz.1.");
```

Comment (Habbie) [reply pdns/recursordist/test-filterpo_cc.cc:826]: committed, they also pass :)

File: pdns/filterpo.cc [hunk 4]
```diff
@@ -697,31 +697,58 @@ DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask& nm)
...
+    //
+    //    If there exists more than one sequence of zero-valued fields of
+    //    identical length, then only the last such sequence is compressed.
+    //    Note that [RFC5952] specifies compressing the first such sequence,
+    //    but our notation here reverses the order of fields, and so must also
+    //    reverse the selection of which zero sequence to compress.
+    //
+    // 'cur.len > best.len' from the original code is replaced by 'cur.len >= best.len', so the last-longest wins.
+
+    struct { int base, len; } best = {-1, 0}, cur = {-1, 0};
```
Comment (omoerbeek) [inline pdns/filterpo.cc:715]: some `size_t` (or `ssize_t`) also for the `i` and you can do away with the casts

**Review by omoerbeek — APPROVED** [review]
Small nit, approved either way

File: pdns/filterpo.cc [hunk 5]
```diff
@@ -697,31 +697,58 @@ DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask& nm)
...
-    while (begin < len) {
-      std::string::size_type end = str.find(":", begin);
-      std::string sub;
-      if (end != string::npos) {
-        sub = str.substr(begin, end - begin);
-      }
-      else {
-        sub = str.substr(begin);
+    static_assert(sizeof(addr.sin6.sin6_addr.s6_addr) == sizeof(uint16_t) * 8);
+    uint16_t *src = (uint16_t*) &addr.sin6.sin6_addr.s6_addr;
```
Comment (rgacogne) [inline pdns/filterpo.cc:701]:
```suggestion
    auto src = reinterpret_cast<const uint16_t*>(&addr.sin6.sin6_addr.s6_addr);
```

File: pdns/filterpo.cc [hunk 6]
```diff
@@ -697,31 +697,58 @@ DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask& nm)
...
+    //    but our notation here reverses the order of fields, and so must also
+    //    reverse the selection of which zero sequence to compress.
+    //
+    // 'cur.len > best.len' from the original code is replaced by 'cur.len >= best.len', so the last-longest wins.
+
+    struct { int base, len; } best = {-1, 0}, cur = {-1, 0};
+
+    for (int i = 0; i < (int)elems.size(); i++) {
+      auto elem = ntohs(src[i]);
+      elems[i] = elem;
```
Comment (rgacogne) [inline pdns/filterpo.cc:719]: Not sure `elem` is really useful here, perhaps we could do `elems[i] = ntohs(src[i]);`?

File: pdns/filterpo.cc [hunk 7]
```diff
@@ -697,31 +697,58 @@ DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask& nm)
...
       }
+    }
 
-      if (end == string::npos) {
-        break;
+    if (best.base != -1 && best.len < 2) {              // if our best run is only one zero long, we do not replace it
+      best.base = -1;
+    }
+    for (int i=0; i < (int)elems.size(); i++) {
+      if (i==best.base) {
```
Comment (rgacogne) [inline pdns/filterpo.cc:746]:
```suggestion
      if (i == best.base) {
```

**Review by rgacogne — APPROVED** [review]
Looks great, nice work! I suggested a couple changes but nothing that really matters.

File: pdns/filterpo.cc (hunk 4 above)
Comment (Habbie) [reply pdns/filterpo.cc:715]: I tried that earlier and got
```
filterpo.cc: In static member function ‘static DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask&)’:
filterpo.cc:717:27: warning: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘long int’} and ‘long unsigned int’ [-Wsign-compare]
     for (ssize_t i = 0; i < elems.size(); i++) {
                         ~~^~~~~~~~~~~~~~
```

Comment (Habbie) [reply pdns/filterpo.cc:715]: The few casts seemed like the best deal :)

File: pdns/filterpo.cc (hunk 6 above)
Comment (Habbie) [reply pdns/filterpo.cc:719]: yep! I even missed a spot where I could have replaced `elems[i]` with `elem`. The separate `elem` var was a leftover from a longer version.

File: pdns/filterpo.cc (hunk 4 above)
Comment (omoerbeek) [reply pdns/filterpo.cc:715]: I was thinking `ssize_t` for the struct members and `size_t` for the loop vars, but oh well.

**Pull request merged**